src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -222,13 +224,57 @@ ucs_status_t uct_dc_mlx5_ep_atomic32_fetch(uct_ep_h ep, uct_atomic_op_t opcode,
                                           remote_addr, rkey, comp);
 }
 
+static ucs_status_t uct_dc_mlx5_iface_fence_pending_cb(uct_pending_req_t *self)
+{
+    uct_rc_iface_fence_req_t *req = ucs_derived_of(self, uct_rc_iface_fence_req_t);
+    uct_dc_mlx5_ep_t *ep          = ucs_derived_of(req->tl_ep, uct_dc_mlx5_ep_t);
+
+    ep->flags |= UCT_DC_MLX5_EP_FLAG_FENCE;
+    ucs_mpool_put(req);
+    return UCS_OK;
+}
+
 ucs_status_t uct_dc_mlx5_ep_fence(uct_ep_h tl_ep, unsigned flags)
 {
     uct_dc_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_dc_mlx5_iface_t);
     uct_dc_mlx5_ep_t *ep       = ucs_derived_of(tl_ep, uct_dc_mlx5_ep_t);
+    uct_rc_iface_fence_req_t *fence_req;
+
+    UCT_TL_EP_STAT_FENCE(ucs_derived_of(tl_ep, uct_base_ep_t));
+
+    if (iface->super.super.config.fence_mode == UCT_RC_FENCE_MODE_NONE) {
+        return UCS_OK;
+    }
+
+    if (!uct_dc_mlx5_iface_has_tx_resources(iface)) {
+        goto no_res;
+    }
+
+    if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
+        if (!uct_dc_mlx5_iface_dci_can_alloc(iface)) {
+            goto no_res; /* waiting for dci */
+        } else {
+            UCT_TL_EP_STAT_FENCE(&ep->super); /* no sends */
+            return UCS_OK;
+        }
+    }
+
+    if (!uct_dc_mlx5_iface_dci_ep_can_send(ep)) {
+        goto no_res; /* cannot send */
+    }
+
+    ep->flags |= UCT_DC_MLX5_EP_FLAG_FENCE;
+    return UCS_OK;
+
+no_res:
+    fence_req = (uct_rc_iface_fence_req_t*)ucs_mpool_get(&iface->super.super.tx.fence_mp);
```

Comment:
maybe unite with rc fence pending add? So that we have code for the same functionality in one place/function